rfd/0154-logging-guidelines.md

```diff
@@ -0,0 +1,204 @@
+by using `fmt.Sprintf` directly it is not recommended. Instead, when converting to `log/slog` structured logging should
+be embraced. The message from above can be translated to carry a bit more context, and the additional fields in the
+message can be moved to the attributes.
+
+> ```go
+> slog.Info("Proxied connection to cluster established.", "cluster_name", clusterName, "remote_address", conn.RemoteAddr())
+>```
+>```bash
+>2023-11-02T15:18:40-04:00 [PROXY]    INFO Proxied connection to cluster established. pid:14968.1 cluster_name:foo remote_address:192.168.1.243  regular/proxy.go:250
+>```
+
+#### WithField(s)
+
+To add structured fields with `logrus` either the `WithField` or `WithFields` methods were used, the only difference
+being the latter took a map of fields instead of a single field.
+
+```go
+logrus.WithField("user", user).Warn("Failed to emit account recovery code used failed event.")
+```
+
+`log/slog` provides similar functionality via the `With` method, though the API differs slightly. Just like the logging
+API, `With` takes a series of key value pairs instead of a struct like `logrus`. The above example can naively be
+converted to the following with `log/slog`.
+
+```go
+slog.With("user", user).Warn("Failed to emit account recovery code used failed event.")
+```
+
+However, instead of creating a clone of the logger with the `user` field added, the desired attributes should be
+passed directly to the `Warn` method instead.
+
+```go
+slog.Warn("Failed to emit account recovery code used failed event.", "user", user)
+```
+
+The main benefit to using `slog.With` is that any attributes provided to it are formatted once instead of each time a
+message is logged. `slog.With` should only be used when enriching a logger with additional information that is used to
+log several messages, be that in a function or a logger that is a struct member.
+
+> `logrus`
+>```go
+>logrusLogger := logrus.WithFields(logrus.Fields{
+>"device_id": ref.DeviceID,
+>"os_type":   ref.OSType,
+>"asset_tag": assetTag,
+>})
+>...
+>logrusLogger.Debug("some message")
+>...
+>logrusLogger.Warn("some other message")
+>```
+>`log/slog`
+>```go
+>slogLogger := slog.With("device_id", ref.DeviceID, "os_type", ref.OSType, "asset_tag", assetTag)
+>...
+>slogLogger.Debug("some message")
+>...
+>slogLogger.Warn("some other message")
+>```
+
+#### WithError
+
+Unlike `logrus`, `log/slog` does not provide any special APIs for dealing with errors. When converting to `log/slog` any
+usages of `WithError` should follow the same advice from the `WithField(s)` section. If the error is only included in a
+single log message, then it should be added as an attribute to the log function. If the error is included in multiple
+messages
+then a logger should be created via `With("error", err)`.
+
+
+> `logrus`
+>```go
+>logrus.WithError(err).Error("Error parsing response.")
+>```
+>`log/slog`
+>```go
+>slog.Error("Error parsing response.", "error", err)
+>```
+
+### Best Practices
```

Comment:
I do personally prefer Attr - only because it reduces the likelihood of "forgetting" to include a key or getting keys and values mixed up - but I don't have a strong preference or feeling that we need to "force" a choice here - other than in cases where performance matters.